### DISCUSSION TITLE
Query contains edge

### DISCUSSION THREAD

**Discussion by dm356** [google/s2geometry#383] [Q&A] [answered]
I like the efficiency of the S2ContainsPointQuery.  Is there a similar option for checking if an edge is contained (line segment between two points)?  What would be the most efficient way to make many of these checks (independently) against a single S2ShapeIndex?

**Comment by smcallis** [comment]
Would you want to check if an edge is contained in a polygon or another line string?

For the former, you could just check that one of the vertices is contained and that the edge doesn't cross any other edges with S2CrossingEdgeQuery, and for the latter just check that both vertices belong to the line string and that they're adjacent.    If you want generic containment capabilities, I think S2BooleanOperation is your best best.

**Reply by dm356** [reply]
Within a polygon.  I tried the S2CrossingEdgeQuery, but it's not responding the same as the S2ContainsPointQuery for points.  Is there a way to check if the edge is fully contained, not just alongside the boundary of the polygon?  This would be similar to the S2VertexModel::OPEN option, if I understand it correctly.

**Comment by smcallis** [comment, marked as answer]
It should give the same answers since it uses the same crossing edge predicates.  To clarify you'll have to do a two step process:

1. Check that the edge doesn't cross any other edges (if this is true, the edge is entirely contained or disjoint from the polygon)
2. Break the tie by checking one of the vertices.

**Reply by dm356** [reply]
Makes sense.  Thanks!
